src/main/java/org/truffleruby/language/SafepointManager.java

```diff
@@ -159,6 +141,46 @@ private SafepointAction step(Node currentNode, boolean isDrivingThread) {
         return deferredAction;
     }
 
+    private void driveArrivalAtPhaser() {
+        int phase = phaser.arrive();
+        long t0 = System.nanoTime();
+        long max = t0 + WAIT_TIME * 1_000_000_000L;
+        int waits = 1;
+        while (true) {
+            try {
+                phaser.awaitAdvanceInterruptibly(phase, 100, TimeUnit.MILLISECONDS);
+                break;
+            } catch (InterruptedException e) {
+                // retry
+            } catch (TimeoutException e) {
+                if (System.nanoTime() >= max) {
+                    Log.LOGGER.severe(String.format("waited %d seconds in the SafepointManager but %d of %d threads did not arrive - a thread is likely making a blocking native call which should use runBlockingSystemCallUntilResult() - check with jstack",
+                            waits * WAIT_TIME, phaser.getUnarrivedParties(), phaser.getRegisteredParties()));
+                    if (waits == 1) {
+                        restoreDefaultInterruptHandler();
+                    }
+                    max += WAIT_TIME * 1_000_000_000L;
+                    waits++;
+                } else {
+                    // Retry interrupting other threads, as they might not have been yet
+                    // in the blocking call when the signal was sent.
+                    interruptOtherThreads();
+                }
+            }
+        }
+    }
+
+    private void restoreDefaultInterruptHandler() {
+        Log.LOGGER.warning("restoring default interrupt handler");
+
+        try {
+            final Signal signal = context.getNativePlatform().getSignalManager().createSignal("INT");
+            context.getNativePlatform().getSignalManager().watchDefaultForSignal(signal);
+        } catch (Throwable t) {
+            Log.LOGGER.warning("failed to restore default interrupt handler");
```

Comment:
Please add `+ "\n" + t` otherwise it becomes very hard to debug.

Reply:
Add it where? This message is only 44 characters. Does it really need a line break in it?

Reply:
Add it at the end of the warning message. Maybe just exception class + message without a \n is enough. At least we can start guessing why it happens.

Reply:
Ah I see what you mean.

Reply:
#488